[Serve] Followup of #42820; Add a DeepSpeed example for HPU (#44317)
ray-project/ray: doc/source/serve/tutorials/intel-habana-gaudi-inference.md

> [Habana Gaudi AI Processors (HPUs)](https://habana.ai) are AI hardware accelerators designed by Habana Labs. For more information, see [Gaudi Architecture](https://docs.habana.ai/en/latest/Gaudi_Overview/index.html) and [Gaudi Developer Docs](https://developer.habana.ai/).

Reviewer: ```suggestion
[Habana Gaudi AI Processors (HPUs)](https://habana.ai) are AI hardware accelerators designed by Habana Labs. See [Gaudi Architecture](https://docs.habana.ai/en/latest/Gaudi_Overview/index.html) and [Gaudi Developer Docs](https://developer.habana.ai/) for more details.
```

> This tutorial has two examples:

Reviewer 2: ```suggestion
This tutorial walks through two examples:
```

> This tutorial walks through two examples:

Reviewer: ```suggestion
This tutorial has two examples:
```

1. Deployment of [Llama2-7b](https://huggingface.co/meta-llama/Llama-2-7b-chat-hf) using a single HPU:

>     * How to load a model in an HPU

Reviewer 2: ```suggestion
    * How to load a model onto an HPU
```

>     * How to load a model onto an HPU

Reviewer: ```suggestion
    * Load a model onto an HPU.
```

>     * How to perform generation on an HPU

Reviewer: ```suggestion
    * Perform generation on an HPU.
```

>     * How to enable HPU Graph optimizations

Reviewer: ```suggestion
    * Enable HPU Graph optimizations.
```

2. Deployment of [Llama2-70b](https://huggingface.co/meta-llama/Llama-2-70b-chat-hf) using multiple HPUs on a single node:

>     * How to initialize distributed backend

Reviewer 2: ```suggestion
    * How to initialize a distributed backend
```

>     * How to initialize a distributed backend

Reviewer: ```suggestion
    * Initialize a distributed backend.
```

>     * How to load a sharded model onto DeepSpeed workers

Reviewer: ```suggestion
    * Load a sharded model onto DeepSpeed workers.
```

>     * How to stream responses from DeepSpeed workers

Reviewer: ```suggestion
    * Stream responses from DeepSpeed workers.
```

> This tutorial helps you serve a large language model on HPUs.

Reviewer: ```suggestion
This tutorial serves a large language model (LLM) on HPUs.
```

> Welcome to try out [llm-on-ray](https://github.com/intel/llm-on-ray), where you can find more examples, and get them deployed in a simpler way.

Reviewer 2: Could you move this to the end of the tutorial in a separate `Next Steps` section? You could say:

```
Check out [llm-on-ray](https://github.com/intel/llm-on-ray) for more ways to customize and deploy LLMs at scale.
```

## Environment setup

> We recommend using a prebuilt container to run these examples. To run a container, you need Docker. See [Install Docker Engine](https://docs.docker.com/engine/install/) for installation instructions.

Reviewer: ```suggestion
Use a prebuilt container to run these examples. To run a container, you need Docker. See [Install Docker Engine](https://docs.docker.com/engine/install/) for installation instructions.
```

```bash
pip install ray[tune,serve]
pip install git+https://github.com/huggingface/optimum-habana.git

> # Replace 1.14.0 with the driver version of the container

Reviewer: ```suggestion
# Replace 1.14.0 with the driver version of the container.
```

pip install git+https://github.com/HabanaAI/DeepSpeed.git@1.14.0

> # Only needed by the DeepSpeed example

Reviewer: ```suggestion
# Only needed by the DeepSpeed example.
```

In this kingdom, there was a beautiful princess named Lily. She was kind, gentle, and had a heart full of laughter. Every day, she would wake up with a smile on her face, ready to face whatever adventures the day might bring.
One day, a wicked sorcerer cast a spell on the kingdom, causing all
```

## Running a sharded model on multiple HPUs

This example shows how to deploy a Llama2-70b model using 8 HPUs orchestrated by DeepSpeed. 

The example requires caching the Llama2-70b model. Run the following Python code in the Habana container to cache the model. 

```python
from huggingface_hub import snapshot_download
snapshot_download(
    "meta-llama/Llama-2-70b-chat-hf",
    # Replace the path if necessary
    cache_dir=os.getenv("TRANSFORMERS_CACHE", None),
    # Specify your Hugging Face token
    token=""
)
```

> In this example, the deployment replica sends prompts to the Deepspeed workers, which are running in Ray actors:

Reviewer 2: ```suggestion
In this example, the deployment replica sends prompts to the DeepSpeed workers, which are running in Ray actors:
```

In this kingdom, there was a beautiful princess named Lily. She was kind, gentle, and had a heart full of laughter. Every day, she would wake up with a smile on her face, ready to face whatever adventures the day might bring.
One day, a wicked sorcerer cast a spell on the kingdom, causing all
```

## Running a sharded model on multiple HPUs

> This example shows how to deploy a Llama2-70b model using 8 HPUs orchestrated by DeepSpeed. 

Reviewer: ```suggestion
This example deploys a Llama2-70b model using 8 HPUs orchestrated by DeepSpeed. 
```

The example requires caching the Llama2-70b model. Run the following Python code in the Habana container to cache the model. 

```python
from huggingface_hub import snapshot_download
snapshot_download(
    "meta-llama/Llama-2-70b-chat-hf",

>     # Replace the path if necessary

Reviewer: ```suggestion
    # Replace the path if necessary.
```

    cache_dir=os.getenv("TRANSFORMERS_CACHE", None),

>     # Specify your Hugging Face token

Reviewer: ```suggestion
    # Specify your Hugging Face token.
```

```text
Once upon a time, there was a young woman named Sophia who lived in a small village nestled in the rolling hills of Tuscany. Sophia was a curious and adventurous soul, always eager to explore the world around her. One day, while wandering through the village, she stumbled upon a hidden path she had never seen before.
The path was overgrown with weeds and vines, and it looked as though it hadn't been traversed in years. But Sophia was intrigued, and she decided to follow it to see where it led. She pushed aside the branches and stepped onto the path
Once upon a time, there was a young woman named Sophia who lived in a small village nestled in the rolling hills of Tuscany. Sophia was a curious and adventurous soul, always eager to explore the world around her. One day, while wandering through the village, she stumbled upon a hidden path she had never seen before.
The path was overgrown with weeds and vines, and it looked as though it hadn't been traversed in years. But Sophia was intrigued, and she decided to follow it to see where it led. She pushed aside the branches and stepped onto the path
```

## Next Steps

> Check out [llm-on-ray](https://github.com/intel/llm-on-ray) for more ways to customize and deploy LLMs at scale.

Reviewer: ```suggestion
See [llm-on-ray](https://github.com/intel/llm-on-ray) for more ways to customize and deploy LLMs at scale.
```
